File: sdk/src/types/analytics.ts
/**
 * Analytics Types for Process Mining SDK
 */

import { HypermediaResponse } from "./common.js";

// =============================================================================
// DASHBOARD
// =============================================================================

export interface DashboardData extends HypermediaResponse {
  logId: string;
  overview: {
    totalCases: number;
    totalEvents: number;
    uniqueActivities: number;
    uniqueResources: number;
    avgCaseDuration: number;
    medianCaseDuration: number;
  };
  topVariants: Array<{
    activities: string[];
    count: number;
    percentage: number;
  }>;
  activityDistribution: Record<string, number>;
  timeDistribution: Array<{
    period: string;
    caseCount: number;
    eventCount: number;
  }>;
}

// =============================================================================
// INSIGHTS
// =============================================================================

export type InsightSeverity = "info" | "warning" | "critical";
export type InsightCategory = "performance" | "conformance" | "resource" | "variant";

export interface ProcessInsight extends HypermediaResponse {
  id: string;
  category: InsightCategory;
  severity: InsightSeverity;
  title: string;
  description: string;
  metric?: number;
  recommendation?: string;
}

// =============================================================================
// ANOMALIES
// =============================================================================

export interface AnomalousCase extends HypermediaResponse {
  caseId: string;
  anomalyType: string;
  score: number;
  details: string;
  activities: string[];
}

export interface AnomalyAnalysis extends HypermediaResponse {
  logId: string;
  totalAnomalies: number;
  anomalyRate: number;
  cases: AnomalousCase[];
}

// =============================================================================
// VARIANT STATS
// =============================================================================

export interface VariantStatistics extends HypermediaResponse {
  logId: string;
  totalVariants: number;
  topVariants: Array<{
    rank: number;
    activities: string[];
    caseCount: number;
    percentage: number;
    avgDuration?: number;
  }>;
  variantDistribution: {
    singleCaseVariants: number;
    multiCaseVariants: number;
    coverageTop5: number;
    coverageTop10: number;
  };
}

// =============================================================================
// RESOURCE STATS
// =============================================================================

export interface ResourceStatistics extends HypermediaResponse {
  logId: string;
  totalResources: number;
  resources: Array<{
    name: string;
    eventCount: number;
    caseCount: number;
    avgEventsPerCase: number;
    activities: string[];
  }>;
}

// =============================================================================
// TIME ANALYSIS
// =============================================================================

export interface TimeAnalysis extends HypermediaResponse {
  logId: string;
  dateRange: {
    start: string;
    end: string;
  };
  casesOverTime: Array<{
    period: string;
    count: number;
  }>;
  avgDurationOverTime: Array<{
    period: string;
    avgDuration: number;
  }>;
  peakHours?: number[];
  peakDays?: string[];
}

// =============================================================================
// PERFORMANCE METRICS (used by AnalyticsClient)
// Aligned with backend schemas.py
// =============================================================================

export interface Bottleneck {
  activity: string;
  avgWaitingTime: number;
  avgServiceTime: number;
  frequency: number;
  isBottleneck: boolean;
  severity: "low" | "medium" | "high";
  impactScore: number;
}

export interface BottleneckListResponse extends HypermediaResponse {
  logId: string;
  bottlenecks: Bottleneck[];
  totalBottlenecks: number;
}

export interface ReworkActivity {
  activity: string;
  reworkCount: number;
  casesWithRework: number;
  reworkPercentage: number;
}

export interface ReworkListResponse extends HypermediaResponse {
  logId: string;
  reworkActivities: ReworkActivity[];
  totalReworkCases: number;
  reworkPercentage: number;
}

export interface ServiceTimeResponse {
  activity: string;
  minSeconds: number;
  maxSeconds: number;
  avgSeconds: number;
  medianSeconds: number;
  stdDevSeconds: number;
}

export interface CycleTimeResponse extends HypermediaResponse {
  logId: string;
  minSeconds: number;
  maxSeconds: number;
  avgSeconds: number;
  medianSeconds: number;
  percentile25: number;
  percentile75: number;
  percentile95: number;
}

export interface ThroughputResponse extends HypermediaResponse {
  logId: string;
  totalCases: number;
  completedCases: number;
  casesPerDay: number;
  casesPerWeek: number;
  casesPerMonth: number;
  timeRangeDays: number;
}

export interface PatternResponse {
  pattern: string;
  frequency: number;
  support: number;
}

/**
 * Performance data used by FE PerformanceTab
 * Matches what getPerformance() returns after transformation
 */
export interface PerformanceData extends HypermediaResponse {
  logId: string;
  cycleTime: CycleTimeResponse;
  throughput: ThroughputResponse;
  topBottlenecks: Array<{
    activity: string;
    avgWaitingTime: number;
    impactScore: number;
  }>;
  reworkSummary: {
    totalReworkCases: number;
    reworkPercentage: number;
  };
}

export interface PerformanceDashboardResponse extends HypermediaResponse {
  logId: string;
  cycleTime: CycleTimeResponse;
  throughput: ThroughputResponse;
  topBottlenecks: Bottleneck[];
  reworkSummary: Record<string, unknown>;
}

File: backend/src/features/process_mining/schemas/analytics.py
"""Analytics schemas - CQRS-compliant response models.

Contains schemas for analytics query responses:
- Bottleneck detection
- Rework analysis
- Cycle time statistics
- Process variants
"""

from pydantic import BaseModel, Field

# =============================================================================
# Bottleneck Analysis
# =============================================================================


class BottleneckResponse(BaseModel):
    """Single bottleneck result."""

    activity: str
    avg_wait_time: float = Field(description="Average wait time in seconds")
    median_wait_time: float = Field(default=0, description="Median wait time in seconds")
    max_wait_time: float = Field(default=0, description="Maximum wait time in seconds")
    count: int = Field(description="Number of occurrences")


class BottleneckListResponse(BaseModel):
    """List of detected bottlenecks."""

    dataset_id: str
    bottlenecks: list[BottleneckResponse]
    total_bottlenecks: int


# =============================================================================
# Rework Analysis
# =============================================================================


class ReworkResponse(BaseModel):
    """Single rework pattern result."""

    activity: str
    repeat_count: int = Field(description="Total repeat occurrences")
    case_count: int = Field(description="Number of cases with this rework")
    percentage: float = Field(description="Percentage of cases with this rework")


class ReworkListResponse(BaseModel):
    """Rework analysis results."""

    dataset_id: str
    rework_activities: list[ReworkResponse]
    total_rework_cases: int
    rework_percentage: float


class ReworkChain(BaseModel):
    """A chain of consecutive rework activities."""

    activity: str
    chain_length: int
    frequency: int
    avg_chain_duration_seconds: float = 0.0
    example_case_ids: list[str] = []


class ReworkChainListResponse(BaseModel):
    """Response for rework chain analysis."""

    dataset_id: str
    chains: list[ReworkChain]
    total_chains: int
    most_problematic_activity: str | None = None
    cases_with_chains: int = 0
    chains_percentage: float = 0.0


# =============================================================================
# Cycle Time
# =============================================================================


class CycleTimeResponse(BaseModel):
    """Cycle time (case duration) statistics."""

    dataset_id: str
    mean_duration: float = Field(description="Mean case duration in seconds")
    median_duration: float = Field(description="Median case duration in seconds")
    min_duration: float = Field(description="Minimum case duration in seconds")
    max_duration: float = Field(description="Maximum case duration in seconds")
    std_deviation: float = Field(default=0, description="Standard deviation in seconds")
    total_cases: int = Field(description="Total number of cases")


# =============================================================================
# Process Variants
# =============================================================================


class VariantResponse(BaseModel):
    """Single process variant."""

    variant_id: int
    activities: list[str] = Field(description="Sequence of activities in this variant")
    case_count: int = Field(description="Number of cases following this variant")
    percentage: float = Field(description="Percentage of total cases")


class VariantListResponse(BaseModel):
    """Process variants analysis results."""

    dataset_id: str
    variants: list[VariantResponse]
    total_variants: int
    total_cases: int


# =============================================================================
# Service Time & Throughput (for backward compatibility)
# =============================================================================


class ServiceTimeResponse(BaseModel):
    """Service time per activity."""

    activity: str
    min_seconds: float
    max_seconds: float
    avg_seconds: float
    median_seconds: float
    std_dev_seconds: float = 0


class ThroughputResponse(BaseModel):
    """Throughput metrics."""

    dataset_id: str
    total_cases: int
    completed_cases: int = 0
    cases_per_day: float = 0
    cases_per_week: float = 0
    cases_per_month: float = 0
    time_range_days: float = 0


class PatternResponse(BaseModel):
    """Frequent pattern result."""

    pattern: list[str]
    support: float
    case_count: int


# =============================================================================
# Performance Dashboard (aggregated view)
# =============================================================================


class PerformanceDashboardResponse(BaseModel):
    """Comprehensive performance dashboard."""

    dataset_id: str
    cycle_time: CycleTimeResponse
    throughput: ThroughputResponse
    top_bottlenecks: list[BottleneckResponse]
    rework_summary: dict
